audio-worklet/noisegate-test.html

```diff
@@ -0,0 +1,432 @@
+        const threshold = toDecibel_(length / 2);
+        let noisegate = new NoiseGate(context, {threshold});
+        let weights = noisegate.computeGain_(envelope);
+        let closedWeights = weights.slice(0, transitionIndex);
+        let openWeights = weights.slice(transitionIndex, length);
+        
+        should(closedWeights, 'Weight array before threshold')
+            .beConstantValueOf(0);
+        should(openWeights, 'Weight array after the threshold')
+            .beConstantValueOf(1);
+        
+        // Verify that the gain computer generates a sequence of increasing or
+        // decreasing weights between 0 and 1 at threshold transitions.
+        // TransitionSteps must be less than half the envelope length so that
+        // both attack and release fit in the evaluation.
+        const transitionSteps = 10;
+        attack = toTime_(transitionSteps, sampleRate);
+        release = attack;
+        noisegate = new NoiseGate(context, {attack, release, threshold});
+        weights = noisegate.computeGain_(envelope);
+
+        let attackWeights = weights.slice(0, transitionSteps);
+        closedWeights = weights.slice(transitionSteps, transitionIndex);
+        let releaseWeights =
+            weights.slice(transitionIndex, transitionIndex + transitionSteps);
+        openWeights =
+            weights.slice(transitionIndex + transitionSteps, length);
+        let expectedAttackWeights = new Float32Array(transitionSteps);
+        let expectedReleaseWeights = new Float32Array(transitionSteps);
+        
+        // Expected weights should align with noise gates weights. Attack
+        // weights increase and release weights decrease.
+        for (let j = 0; j < transitionSteps; j++) {
+          // The first attack weight is 1 - (1 / transitionSteps).
+          expectedAttackWeights[j] = 1 - ((j + 1) / transitionSteps);
+          
+          // Similarly, the first release weight is 1 / transitionSteps.
+          expectedReleaseWeights[j] =  (j + 1) / transitionSteps;
+        }
+
+        // Allow for a small error (set experimentally) since the exact change
+        // in weight is computed differently here and in noisegate.computeGain.
+        const errorThreshold = 1e-6;
+        should(attackWeights, 'Attacking weights')
+            .beCloseToArray(expectedAttackWeights, {absoluteThreshold: 1e-6});
+        should(releaseWeights, 'Releasing weights')
+            .beCloseToArray(expectedReleaseWeights, {absoluteThreshold: 1e-6});
+        should(closedWeights, 'Weight array before threshold')
+            .beConstantValueOf(0);
+        should(openWeights, 'Weight array after the threshold')
+            .beConstantValueOf(1);
+        task.done();
+      });
+
+    audit.run();
+
+    // End of test cases. Testing utility functions below!
+
+    function getNumStepsFromZeroUpTo_(threshold, alpha) {
+      return Math.floor(Math.log10(1 - (Math.pow(threshold, 2) / 2)) /
```

Comment:
Some comments needed here so this formula doesn't look like it was pulled out of thin air!